tool/src/org/antlr/v4/tool/GrammarParserInterpreter.java

```diff
@@ -0,0 +1,318 @@
+package org.antlr.v4.tool;
+
+import org.antlr.v4.runtime.BailErrorStrategy;
+import org.antlr.v4.runtime.InterpreterRuleContext;
+import org.antlr.v4.runtime.Parser;
+import org.antlr.v4.runtime.ParserInterpreter;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.TokenStream;
+import org.antlr.v4.runtime.Vocabulary;
+import org.antlr.v4.runtime.atn.ATN;
+import org.antlr.v4.runtime.atn.ATNDeserializer;
+import org.antlr.v4.runtime.atn.ATNSerializer;
+import org.antlr.v4.runtime.atn.ATNState;
+import org.antlr.v4.runtime.atn.DecisionState;
+import org.antlr.v4.runtime.atn.PredictionMode;
+import org.antlr.v4.runtime.atn.RuleStartState;
+import org.antlr.v4.runtime.atn.StarLoopEntryState;
+import org.antlr.v4.runtime.tree.Trees;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** A heavier weight {@link ParserInterpreter} that creates parse trees
+ *  that track alternative numbers for subtree roots.
+ *
+ * @since 4.5.1
+ *
+ */
+public class GrammarParserInterpreter extends ParserInterpreter {
+	/** The grammar associated with this interpreter. Unlike the
+	 *  {@link ParserInterpreter} from the standard distribution,
+	 *  this can reference Grammar, which is in the tools area not
+	 *  purely runtime.
+	 */
+	protected final Grammar g;
+
+	protected BitSet decisionStatesThatSetOuterAltNumInContext;
+
+	/** Cache {@link LeftRecursiveRule#getPrimaryAlts()} and
+	 *  {@link LeftRecursiveRule#getRecursiveOpAlts()} for states in
+	 *  {@link #decisionStatesThatSetOuterAltNumInContext}.
+	 */
+	protected final Map<DecisionState, int[]> stateToAltsMap = new HashMap<DecisionState, int[]>();
```

Comment:
:bulb: If performance accessing this map matters, this would be much faster as an `ArrayList<int[]>`, indexed by decision number.